File: ios/RNGoogleIMAConverters.m
@import GoogleInteractiveMediaAds;

id orNull(id value) {
    return value != nil ? value : [NSNull null];
}

NSMutableArray* convertArray(NSArray* source) {
    NSMutableArray *target = [[NSMutableArray alloc] init];
    if (source != nil) {
        for (int i = 0; i < source.count; ++i) {
            [target addObject:[source objectAtIndex:i]];
        }
    }
    return target;
}

NSDictionary* convertAdsManager(IMAAdsManager* adsManager)
{
    return @{
             @"adCuePoints": convertArray(adsManager.adCuePoints),
             @"adPlaybackInfo": @{
                     @"currentMediaTime": [NSNumber numberWithDouble:adsManager.adPlaybackInfo.currentMediaTime],
                     @"totalMediaTime": [NSNumber numberWithDouble:adsManager.adPlaybackInfo.totalMediaTime],
                     @"bufferedMediaTime": [NSNumber numberWithDouble:adsManager.adPlaybackInfo.bufferedMediaTime],
                     @"playing": [NSNumber numberWithBool:adsManager.adPlaybackInfo.playing],
                     }
             };
}

NSDictionary* convertStreamManager(IMAStreamManager* streamManager)
{
    return @{@"streamId": streamManager.streamId};
}

NSDictionary* convertAdError(IMAAdError* adError)
{
    return @{
             @"type": [NSNumber numberWithLongLong:adError.type],
             @"code": [NSNumber numberWithLongLong:adError.code],
             @"message": adError.message,
             };
}

NSDictionary* convertAdLoadingErrorData(IMAAdLoadingErrorData* adErrorData)
{
    return @{
             @"adError": convertAdError(adErrorData.adError),
             };
}

NSDictionary* convertAdsLoadedData(IMAAdsLoadedData* adsLoadedData)
{
    return @{
             @"adsManager": convertAdsManager(adsLoadedData.adsManager),
             @"streamManager": convertStreamManager(adsLoadedData.streamManager),
             };
}

NSDictionary* convertAdPodInfo(IMAAdPodInfo* adPodInfo)
{
    NSDictionary* adPodInfoDictionary = nil;
    if (adPodInfo != nil) {
        adPodInfoDictionary = @{
                                @"totalAds": [NSNumber numberWithLong:adPodInfo.totalAds],
                                @"adPosition": [NSNumber numberWithLong:adPodInfo.adPosition],
                                @"isBumper": [NSNumber numberWithLong:adPodInfo.isBumper],
                                @"podIndex": [NSNumber numberWithLong:adPodInfo.podIndex],
                                @"timeOffset": [NSNumber numberWithDouble:adPodInfo.timeOffset],
                                @"maxDuration": [NSNumber numberWithDouble:adPodInfo.maxDuration],
                                };
    }
    return adPodInfoDictionary;
}

NSDictionary* convertAd(IMAAd* ad) {
    NSDictionary* adDictionary = nil;
    @try {
        if (ad != nil) {
            adDictionary =@{
                @"adId": orNull(ad.adId),
                @"adTitle": orNull(ad.adTitle),
                @"adDescription": orNull(ad.adDescription),
                @"adSystem": orNull(ad.adSystem),
                @"contentType": orNull(ad.contentType),
                @"duration": [NSNumber numberWithDouble:ad.duration],
                @"uiElements": convertArray(ad.uiElements),
                @"width": [NSNumber numberWithInteger:ad.width],
                @"height": [NSNumber numberWithInteger:ad.height],
                @"vastMediaWidth": [NSNumber numberWithInteger:ad.VASTMediaWidth],
                @"vastMediaHeight": [NSNumber numberWithInteger:ad.VASTMediaHeight],
                @"vastMediaBitrate": [NSNumber numberWithInteger:ad.VASTMediaBitrate],
                @"linear": [NSNumber numberWithBool:ad.linear],
                @"skippable": [NSNumber numberWithBool:ad.skippable],
                @"skipTimeOffset": [NSNumber numberWithDouble:ad.skipTimeOffset],
                @"adPodInfo": orNull(convertAdPodInfo(ad.adPodInfo)),
                @"traffickingParameters": orNull(ad.traffickingParameters),
                @"creativeID": orNull(ad.creativeID),
                @"universalAdIdValue": orNull(ad.universalAdIdValue),
                @"universalAdIdRegistry": orNull(ad.universalAdIdRegistry),
                @"advertiserName": orNull(ad.advertiserName),
                @"surveyURL": orNull(ad.surveyURL),
                @"dealID": orNull(ad.dealID),
                @"wrapperAdIDs": convertArray(ad.wrapperAdIDs),
                @"wrapperSystems": convertArray(ad.wrapperSystems),
            };
        }
    } @catch (NSException *exception) {
    } @finally {
        return adDictionary;
    }
}

NSDictionary* convertAdCuepoint(IMACuepoint* cuepoint) {
    NSDictionary* cuepointDictionary;
    
    if (cuepoint != nil) {
        cuepointDictionary = @{
                               @"startTime": [NSNumber numberWithDouble:cuepoint.startTime],
                               @"endTime": [NSNumber numberWithDouble:cuepoint.endTime],
                               @"played": [NSNumber numberWithBool:cuepoint.played],
                               };
    }
    
    return cuepointDictionary;
}

NSDictionary* converAdData(NSDictionary* adData) {
    if (adData) {
        NSMutableDictionary* adDataDictionary = [[NSMutableDictionary alloc] init];
        NSArray<NSString*>* keys = [adData allKeys];
        for (int i = 0; i < keys.count; i++)
        {
            NSString* key = [keys objectAtIndex:i];
            if ([key isEqualToString:@"cuepoints"]) {
                NSArray<IMACuepoint*>* cuepointsArray = [adData objectForKey:key];
                NSMutableArray<NSDictionary*>* cuepointsMutable = [[NSMutableArray alloc] init];
                for (int j = 0; j < cuepointsArray.count; j++)
                {
                    [cuepointsMutable addObject:orNull(convertAdCuepoint([cuepointsArray objectAtIndex:j]))];
                }
                [adDataDictionary setValue:cuepointsMutable forKey:key];
            }
        }
        return adDataDictionary;
    }
    return nil;
}

NSDictionary* convertAdEvent(IMAAdEvent* event) {
    NSDictionary* adEventDictionary = nil;
    if (event != nil) {
        NSString* adEventTypeString = nil;
        switch (event.type) {
            case kIMAAdEvent_STREAM_LOADED: {
                adEventTypeString = @"STREAM_LOADED";
                break;
            }
            case kIMAAdEvent_STREAM_STARTED: {
                adEventTypeString = @"STREAM_STARTED";
                break;
            }
            case kIMAAdEvent_AD_BREAK_READY: {
                adEventTypeString = @"AD_BREAK_READY";
                break;
            }
            case kIMAAdEvent_AD_BREAK_STARTED: {
                adEventTypeString = @"AD_BREAK_STARTED";
                break;
            }
            case kIMAAdEvent_AD_BREAK_ENDED: {
                adEventTypeString = @"AD_BREAK_ENDED";
                break;
            }
            case kIMAAdEvent_AD_PERIOD_STARTED: {
                adEventTypeString = @"AD_PERIOD_STARTED";
                break;
            }
            case kIMAAdEvent_AD_PERIOD_ENDED: {
                adEventTypeString = @"AD_PERIOD_ENDED";
                break;
            }
            case kIMAAdEvent_ALL_ADS_COMPLETED: {
                adEventTypeString = @"ALL_ADS_COMPLETED";
                break;
            }
            case kIMAAdEvent_LOADED: {
                adEventTypeString = @"LOADED";
                break;
            }
            case kIMAAdEvent_STARTED: {
                adEventTypeString = @"STARTED";
                break;
            }
            case kIMAAdEvent_CLICKED: {
                adEventTypeString = @"CLICKED";
                break;
            }
            case kIMAAdEvent_TAPPED: {
                adEventTypeString = @"TAPPED";
                break;
            }
            case kIMAAdEvent_CUEPOINTS_CHANGED: {
                adEventTypeString = @"CUEPOINTS_CHANGED";
                break;
            }
            case kIMAAdEvent_LOG: {
                adEventTypeString = @"LOG";
                break;
            }
            case kIMAAdEvent_PAUSE: {
                adEventTypeString = @"PAUSE";
                break;
            }
            case kIMAAdEvent_RESUME: {
                adEventTypeString = @"RESUME";
                break;
            }
            case kIMAAdEvent_SKIPPED: {
                adEventTypeString = @"SKIPPED";
                break;
            }
            case kIMAAdEvent_FIRST_QUARTILE: {
                adEventTypeString = @"FIRST_QUARTILE";
                break;
            }
            case kIMAAdEvent_MIDPOINT: {
                adEventTypeString = @"MIDPOINT";
                break;
            }
                /**
                 *  Third quartile of a linear ad was reached.
                 */
            case kIMAAdEvent_THIRD_QUARTILE: {
                adEventTypeString = @"THIRD_QUARTILE";
                break;
            }
                /**
                 *  Single ad has finished.
                 */
            case kIMAAdEvent_COMPLETE: {
                adEventTypeString = @"COMPLETE";
                break;
            }
                
            default:
                break;
        }
        adEventDictionary = @{
                              @"type": adEventTypeString,
                              @"typeString": event.typeString,
                              @"ad": orNull(convertAd(event.ad)),
                              @"adData": orNull(converAdData(event.adData)),
                              };
    }
    
    return adEventDictionary;
}

File: android/src/main/java/com/reactlibrary/googleIMA/RNGoogleIMAConverters.java
package com.reactlibrary.googleIMA;

import android.util.Log;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeArray;
import com.facebook.react.bridge.WritableNativeMap;
import com.google.ads.interactivemedia.v3.api.Ad;
import com.google.ads.interactivemedia.v3.api.AdError;
import com.google.ads.interactivemedia.v3.api.AdErrorEvent;
import com.google.ads.interactivemedia.v3.api.AdEvent;
import com.google.ads.interactivemedia.v3.api.AdPodInfo;
import com.google.ads.interactivemedia.v3.api.AdsManager;
import com.google.ads.interactivemedia.v3.api.AdsManagerLoadedEvent;
import com.google.ads.interactivemedia.v3.api.CuePoint;
import com.google.ads.interactivemedia.v3.api.StreamManager;
import com.google.ads.interactivemedia.v3.api.UiElement;

import java.util.List;
import java.util.Map;

public class RNGoogleIMAConverters {
  static WritableMap convertAdsLoadedData(AdsManagerLoadedEvent adsLoadedData) {
    WritableMap data = Arguments.createMap();
    data.putMap("adsManager", convertAdsManager(adsLoadedData.getAdsManager()));
    data.putMap("streamManager", convertStreamManager(adsLoadedData.getStreamManager()));
    WritableNativeMap adsLoadedDataDictionary = new WritableNativeMap();
    adsLoadedDataDictionary.putMap("adsLoadedData", data);
    return adsLoadedDataDictionary;
  }

  static WritableMap convertAdsManager(AdsManager adsManager) {
    if (adsManager == null) {
      return null;
    }
    WritableMap data = Arguments.createMap();

    data.putArray("adCuePoints", convertArray(adsManager.getAdCuePoints().toArray(new Float[0])));

    WritableNativeMap adPlaybackInfo = new WritableNativeMap();
    adPlaybackInfo.putInt("currentMediaTime", (int) adsManager.getAdProgress().getCurrentTimeMs());
    adPlaybackInfo.putInt("totalMediaTime", (int) adsManager.getAdProgress().getDurationMs());
    adPlaybackInfo.putInt("bufferedMediaTime", 0);
    adPlaybackInfo.putBoolean("playing", true);
    data.putMap("adPlaybackInfo", adPlaybackInfo);

    return data;
  }

  static WritableMap convertStreamManager(StreamManager streamManager) {
    WritableMap data = Arguments.createMap();

    data.putString("streamId", streamManager.getStreamId());

    return data;
  }

  static WritableArray convertArray(Object[] items) {
    WritableArray writableArray = Arguments.createArray();
    for (Object item : items) {
      if (item.getClass().equals(String.class)) {
        writableArray.pushString((String) item);
      } else if (item.getClass().equals(String.class)) {
        writableArray.pushString((String) item);
      } else if (item.getClass().equals(UiElement.class)) {
        writableArray.pushString(((UiElement) item).getName());
      }

    }
    return writableArray;
  }

  public static WritableMap convertAdEvent(AdEvent adEvent, List<CuePoint> cuepoints) {
    if (adEvent != null) {
      String adEventTypeString = adEvent.getType().toString();

      if (adEvent.getType() == AdEvent.AdEventType.COMPLETED) {
        adEventTypeString = "COMPLETE";
      }

      WritableNativeMap adEventContentsDictionary = new WritableNativeMap();
      adEventContentsDictionary.putString("type", adEventTypeString);
      adEventContentsDictionary.putString("typeString", adEvent.getType().name());
      adEventContentsDictionary.putMap("ad", convertAd(adEvent.getAd()));
      adEventContentsDictionary.putMap("adData", convertAdData(adEvent.getAdData(), cuepoints));

      WritableNativeMap adEventDictionary = new WritableNativeMap();
      adEventDictionary.putMap("adEvent", adEventContentsDictionary);

      return adEventDictionary;
    }
    return null;
  }

  public static WritableMap convertAdData(Map<String, String> adData, List<CuePoint> cuepoints) {
    WritableMap adDataDictionary = Arguments.createMap();
    if (cuepoints != null) {
      WritableArray cuepointsList = Arguments.createArray();
      for (CuePoint cuePoint : cuepoints) {
        cuepointsList.pushMap(convertAdCuepoint(cuePoint));
      }
      adDataDictionary.putArray("cuepoints", cuepointsList);
    }
    if (adData != null) {
      String[] keys = (String[]) adData.keySet().toArray();
//      for (String key : keys) {
//        if (key.equals("cuepoints")) {
//          adDataDictionary.putString(key, adData.get(key));
////                    adDataDictionary.putArray(key, convertArray());
//        }
//      }
//            for (int i = 0; i < keys.length; i++)
//            {
//                String key = keys[i];
//                if (key.equals("cuepoints")) {
//
//                }
////                if ([key isEqualToString:@"cuepoints"]) {
////                NSArray<IMACuepoint*>* cuepointsArray = [adData objectForKey:key];
////                NSMutableArray<NSDictionary*>* cuepointsMutable = [[NSMutableArray alloc] init];
////                for (int j = 0; j < cuepointsArray.count; j++)
////                {
////                    [cuepointsMutable addObject:orNull(convertAdCuepoint([cuepointsArray objectAtIndex:j]))];
////                }
////                [adDataDictionary setValue:cuepointsMutable forKey:key];
//
//            }
    }
    return adDataDictionary;
  }


  public static WritableMap convertAdCuepoint(CuePoint cuepoint) {
    WritableMap adDataDictionary = Arguments.createMap();

    if (cuepoint != null) {
      adDataDictionary.putDouble("startTime", cuepoint.getStartTime());
      adDataDictionary.putDouble("endTime", cuepoint.getEndTime());
      adDataDictionary.putBoolean("played", cuepoint.isPlayed());
    }

    return adDataDictionary;
  }

  public static WritableMap convertAd(Ad ad) {
    WritableMap adDictionary = Arguments.createMap();
    try {
      if (ad != null) {
        adDictionary.putString("adId", ad.getAdId());
        adDictionary.putString("adTitle", ad.getTitle());
        adDictionary.putString("adDescription", ad.getDescription());
        adDictionary.putString("adSystem", ad.getAdSystem());
        adDictionary.putString("contentType", ad.getContentType());
        adDictionary.putDouble("duration", ad.getDuration());
        adDictionary.putArray("uiElements", convertArray(ad.getUiElements().toArray(new UiElement[0])));
        adDictionary.putInt("width", ad.getWidth());
        adDictionary.putInt("height", ad.getHeight());
        adDictionary.putInt("vastMediaWidth", ad.getVastMediaWidth());
        adDictionary.putInt("vastMediaHeight", ad.getVastMediaHeight());
        adDictionary.putInt("vastMediaBitrate", ad.getVastMediaWidth());
        adDictionary.putBoolean("linear", ad.isLinear());
        adDictionary.putBoolean("skippable", ad.isSkippable());
        adDictionary.putDouble("skipTimeOffset", ad.getSkipTimeOffset());
        adDictionary.putMap("adPodInfo", convertAdPodInfo(ad.getAdPodInfo()));
        adDictionary.putString("traffickingParameters", ad.getTraffickingParameters());
        adDictionary.putString("creativeID", ad.getCreativeId());
        adDictionary.putArray("universalAdIds", convertArray(ad.getUniversalAdIds()));
//                adDictionary.putString("universalAdIdRegistry", ad.getUniversalAdIdRegistry());
        adDictionary.putString("advertiserName", ad.getAdvertiserName());
        adDictionary.putString("surveyURL", ad.getSurveyUrl());
        adDictionary.putString("dealID", ad.getDealId());
        adDictionary.putArray("wrapperAdIDs", convertArray(ad.getAdWrapperIds()));
        adDictionary.putArray("wrapperSystems", convertArray(ad.getAdWrapperSystems()));
      }
    } catch (Exception e) {
      System.out.print(e.getMessage());
    }
    return adDictionary;
  }

  public static WritableMap convertAdPodInfo(AdPodInfo adPodInfo) {
    WritableNativeMap adPodInfoDictionary = new WritableNativeMap();
    if (adPodInfo != null) {
      adPodInfoDictionary.putInt("totalAds", adPodInfo.getTotalAds());
      adPodInfoDictionary.putInt("adPosition", adPodInfo.getAdPosition());
      adPodInfoDictionary.putBoolean("isBumper", adPodInfo.isBumper());
      adPodInfoDictionary.putInt("podIndex", adPodInfo.getPodIndex());
      adPodInfoDictionary.putDouble("timeOffset", adPodInfo.getTimeOffset());
      adPodInfoDictionary.putDouble("maxDuration", adPodInfo.getMaxDuration());
    }
    return adPodInfoDictionary;
  }

  public static WritableMap convertAdErrorEvent(AdErrorEvent adErrorEvent) {
    WritableMap adPodInfoDictionary = Arguments.createMap();
    if (adErrorEvent != null) {
      AdError adError = adErrorEvent.getError();
      if (adError != null) {
        adPodInfoDictionary.putString("type", adError.getErrorType().name());
        adPodInfoDictionary.putString("code", adError.getErrorCode().name());
        adPodInfoDictionary.putString("message", adError.getMessage());
      }
    }
    return adPodInfoDictionary;
  }


}
